src/engine/GroupByImpl.cpp

```diff
@@ -31,6 +31,212 @@
+        singleIdTable_(singleIdTable),
+        inWidth_(subTreeResultWidth),
+        resultTable_(parent->getResultWidth(),
+                     parent->getExecutionContext()->getAllocator()) {
+    AD_CONTRACT_CHECK(inWidth_ == IN_WIDTH || IN_WIDTH == 0);
+    AD_CONTRACT_CHECK(parent_ != nullptr);
+  }
+
+  void onBlockChange(size_t blockStart, size_t blockEnd,
+                     sparqlExpression::EvaluationContext& evaluationContext) {
+    if (groupSplitAcrossTables_) {
+      lazyGroupBy_->processBlock(evaluationContext, blockStart, blockEnd);
+      lazyGroupBy_->commitRow(resultTable_, evaluationContext,
+                              currentGroupBlock_);
+      groupSplitAcrossTables_ = false;
+    } else {
+      // This processes the whole block in batches if possible
+      IdTableStatic<OUT_WIDTH> table =
+          std::move(resultTable_).toStatic<OUT_WIDTH>();
+      parent_->processBlock<OUT_WIDTH>(table, aggregates_, evaluationContext,
+                                       blockStart, blockEnd,
+                                       &currentLocalVocab_, groupByCols_);
+      resultTable_ = std::move(table).toDynamic();
+    }
+  }
+
+  auto process(Result::IdTableVocabPair& pair) {
+    using LoopControl = ad_utility::LoopControl<Result::IdTableVocabPair>;
+    auto& idTable = pair.idTable_;
+    if (idTable.empty()) {
+      return LoopControl::makeContinue();
+    }
+
+    AD_CORRECTNESS_CHECK(idTable.numColumns() == inWidth_);
+    parent_->checkCancellation();
+    storedLocalVocabs_.emplace_back(std::move(pair.localVocab_));
+
+    if (currentGroupBlock_.empty()) {
+      for (size_t col : groupByCols_) {
+        currentGroupBlock_.emplace_back(col, idTable(0, col));
+      }
+    }
+
+    sparqlExpression::EvaluationContext evaluationContext =
+        parent_->createEvaluationContext(currentLocalVocab_, idTable);
+
+    size_t lastBlockStart = parent_->searchBlockBoundaries(
+        [this, &evaluationContext](size_t a, size_t b) {
+          onBlockChange(a, b, evaluationContext);
+        },
+        idTable.asStaticView<IN_WIDTH>(), currentGroupBlock_);
+    groupSplitAcrossTables_ = true;
+    lazyGroupBy_->processBlock(evaluationContext, lastBlockStart,
+                               idTable.size());
+    if (!singleIdTable_ && !resultTable_.empty()) {
+      currentLocalVocab_.mergeWith(storedLocalVocabs_);
+      auto result = Result::IdTableVocabPair{std::move(resultTable_),
+                                             std::move(currentLocalVocab_)};
+      // Keep last local vocab for next commit.
+      currentLocalVocab_ = std::move(storedLocalVocabs_.back());
+      storedLocalVocabs_.clear();
+
+      return LoopControl::yieldValue(std::move(result));
+    }
+
+    return LoopControl::makeContinue();
+  }
+
+  void buildFinalValue() {
+    using namespace ad_utility;
+    using LazyResult = Result::LazyResult;
+    using IdTableVocabPair = Result::IdTableVocabPair;
```

Comment:
Those can be at class or namespace scope respectively, then the functions are a bit more concise.